### PR TITLE
[cargo x] x now shows uses what it's doing in log lines

### PR DESCRIPTION
**Pull request by rexhoffman**
## Motivation

x should show users what commands it's executing, whether they fail or not, and how long they take.
Just what I created for myself a while back, cleaned up for other's given @metajack's pr will make x run some highly configured commands. 

Some extensions that I think we need, but don't want to block on:
* Listeners to capture this data, one to display it inline, and another to capture and write to file/report at end.
* environment variable reports/mutators.
* retry on failed tests -- but add info to listeners.
* capture all tests and all tests results in listeners.
* strict control over environment variables -- show what matters, restrict what doesn't.

### Have you read the [Contributing Guidelines on pull requests](https://github.com/libra/libra/blob/master/CONTRIBUTING.md#pull-requests)?

Yes

## Test Plan

cargo xtest, cargo xtest --unit, cargo xfmt

## Related PRs

Semi related: 
https://github.com/libra/libra/pull/3134

### CODE REVIEW

**Comment by rexhoffman** [comment]
For reference, this is what output looks like, not a huge fan of each command part being quoted, but it doesn't get in the way of cut/pasting the commands.

Screen Shot 2020-03-29 at 10 12 54 PM

File: x/src/cargo.rs [hunk 1]
```diff
@@ -86,19 +88,35 @@ impl Cargo {
...
+        }
+        let now = Instant::now();
         let output = self.inner.output()?;
+        // once all the command has been executed we log it's success or failure.
+        if log {
+            let status = if output.status.success() {
+                "Completed"
+            } else {
+                "Failed"
+            };
```
Comment (rexhoffman) [inline x/src/cargo.rs]: If their's a better way to do this, please let me know, not a fan.

File: x/src/cargo.rs [hunk 2]
```diff
@@ -86,19 +88,35 @@ impl Cargo {
...
+            let elapsed = now.elapsed().as_millis();
+            info!("{} in {} millis: {:?}", status, elapsed, &self.inner);
```
Comment (metajack) [inline x/src/cargo.rs]: Nit: I think `{} in {}ms` would be better.

File: x/src/main.rs [hunk 3]
```diff
@@ -46,6 +47,11 @@ enum Command {
 }
 
 fn main() -> Result<()> {
+    let env = Env::default()
+        .filter_or("MY_LOG_LEVEL", "info")
```
Comment (metajack) [inline x/src/main.rs]: Why not let RUST_LOG do this as usual?

File: .gitignore [hunk 4]
```diff
@@ -38,5 +38,8 @@ Temporary Items
 # Generated VM config in vm-genesis
 language/tools/vm-genesis/genesis/vm_config.toml
 
+#vscode config
+.vscode/
+
```
Comment (bmwill) [inline .gitignore]: Seems unrelated to this change, maybe move this into a separate commit?

File: x/src/cargo.rs [hunk 5]
```diff
@@ -86,19 +88,35 @@ impl Cargo {
...
 
-    fn do_run(&mut self) -> Result<Output> {
+    fn do_run(&mut self, log: bool) -> Result<Output> {
         if !self.pass_through_args.is_empty() {
             self.inner.arg("--").args(&self.pass_through_args);
         }
+        // once all the arguments are added to the command we can log it.
+        if log {
+            info!("Executing: {:?}", &self.inner);
+        }
```
Comment (bmwill) [inline x/src/cargo.rs]: I don't think its worth piping through if we should log or not, remove the bool that you're feeding through and just always log and let the user's set RUST_LOG or a top level command switch that just doesn't enable env_logger

File: x/src/cargo.rs [hunk 6]
```diff
@@ -86,19 +88,35 @@ impl Cargo {
...
+        // once all the command has been executed we log it's success or failure.
+        if log {
+            let status = if output.status.success() {
+                "Completed"
+            } else {
+                "Failed"
+            };
+            let elapsed = now.elapsed().as_millis();
+            info!("{} in {} millis: {:?}", status, elapsed, &self.inner);
+        }
```
Comment (bmwill) [inline x/src/cargo.rs]:
```suggestion

            let status = if output.status.success() {
                "Completed"
            } else {
                "Failed"
            };
            let elapsed = now.elapsed().as_millis();
            info!("{} in {}ms: {:?}", if output.status.success() { "Completed" } else { "Failed" }, now.elapsed().as_millis(), &self.inner);
```

Maybe if you didn't want things separated out

Comment (bmwill) [reply x/src/cargo.rs]: Welp suggestions suck because now I can't edit it and remove the status var or format things

File: x/src/cargo.rs [hunk 7]
```diff
@@ -86,19 +88,35 @@ impl Cargo {
 
     pub fn run(&mut self) -> Result<()> {
         self.inner.stdout(Stdio::inherit()).stderr(Stdio::inherit());
-        self.do_run().map(|_| ())
+        self.do_run(true).map(|_| ())
```
Comment (bmwill) [inline x/src/cargo.rs]: No need for this

File: x/src/main.rs [hunk 8]
```diff
@@ -46,6 +47,11 @@ enum Command {
...
+        .write_style_or("MY_LOG_STYLE", "always");
+    env_logger::init_from_env(env);
+
```
Comment (bmwill) [inline x/src/main.rs]:
```suggestion
    env_logger::from_env(env_logger::Env::default().default_filter_or("info")).init();
```

If you really, really want you can add a `-q` flag to `x` and just not enable the logger if that flag is set

File: x/src/cargo.rs (hunk 5 above)
Comment (rexhoffman) [reply x/src/cargo.rs]: The reason is to prevent calls to pkgid from logging, it's of no value to the user.

File: x/src/main.rs (hunk 8 above)
Comment (rexhoffman) [reply x/src/main.rs]: I discovered this locally as well, as I wanted to default to local, my current approach has a special formatter to make the logging not as visually jarring, and I _think_ should honor RUST_LOG as well (at least for the level).   Maybe there's a way to pull in cargo's default logging style, but didn't see it.

Comment (rexhoffman) [reply x/src/main.rs]: actually you were further ahead,  cloning the local env seemed to be the only way to honor configs, or override with info.   I guess I better get used to this rust with immutable objects.

File: x/src/cargo.rs (hunk 5 above)
Comment (metajack) [reply x/src/cargo.rs]: Maybe inspect the subcommand instead?

Comment (bmwill) [reply x/src/cargo.rs]: I'm not sure I follow, could you provide a little more context?

File: x/src/cargo.rs (hunk 2 above)
Comment (rexhoffman) [reply x/src/cargo.rs]: done

File: x/src/main.rs (hunk 3 above)
Comment (rexhoffman) [reply x/src/main.rs]: supported now :D

File: .gitignore (hunk 4 above)
Comment (rexhoffman) [reply .gitignore]: removed

File: x/src/cargo.rs (hunk 5 above)
Comment (rexhoffman) [reply x/src/cargo.rs]: I'm going to be calling grcov, and not strictly cargo commands soon, so don't want to start parsing the command, at least without capturing all this in the type info first.

File: x/src/cargo.rs (hunk 7 above)
Comment (rexhoffman) [reply x/src/cargo.rs]: some way to default the parameter to true?

**Comment by rexhoffman** [comment]
Oh better output now, and RUST_LOG can turn it off:   
Screen Shot 2020-03-30 at 11 07 25 AM

**Comment by rexhoffman** [comment]
@bmwill @metajack thanks for the feedback, helps me get in to more of a rust mindset and learn these std libs.
